clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ServiceRequest.java

```diff
@@ -71,4 +75,29 @@ public void receiveResponse(Object value, Address sender, boolean exceptional) {
     public boolean completeExceptionally(Throwable exception) {
         return super.completeExceptionally((exception instanceof SuspectedException) ? new CancellationException() : exception);
     }
+
+    @Override
+    public T get() throws InterruptedException, ExecutionException {
+        try {
+            // Wait at most for the configured timeout
+            // If the message was dropped by the receiver, this would otherwise block forever
+            return super.get(super.options.timeout(), TimeUnit.MILLISECONDS);
+        } catch (TimeoutException e) {
+            // Auto-cancel on timeout
+            this.cancel(true);
+            throw new CancellationException(e.getLocalizedMessage());
+        }
+    }
+
+    @Override
+    public T join() {
+        try {
+            return this.get();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new CompletionException(e);
+        } catch (ExecutionException e) {
+            throw new CompletionException(e.getCause());
+        }
```

Comment:
Is letting the CancellationException from L88 propagate unchanged ok?

I have no opinion on this; I'm just checking in case you hadn't noticed that would propagate.

Reply:
Yes - we want to propagate CancellationExceptions.